src/jsMain/kotlin/net/kyori/adventure/webui/js/StyleButtons.kt

```diff
@@ -0,0 +1,206 @@
+import org.w3c.dom.asList
+
+// https://iro.js.org/colorPicker_api.html
+private external interface ColorPicker {
+    // https://iro.js.org/guide.html#color-picker-events
+    fun <T> on(event: String, callback: (T) -> Unit)
+    var color: Color
+}
+
+// https://iro.js.org/color_api.html
+private external interface Color {
+    var hexString: String
+}
+
+// Copied out of adventure; can't use adventure api directly because this is javascript
+private val NAMED_COLORS = mapOf(
+    "#000000" to "black",
+    "#0000aa" to "dark_blue",
+    "#00aa00" to "dark_green",
+    "#00aaaa" to "dark_aqua",
+    "#aa0000" to "dark_red",
+    "#aa00aa" to "dark_purple",
+    "#ffaa00" to "gold",
+    "#aaaaaa" to "gray",
+    "#555555" to "dark_gray",
+    "#5555ff" to "blue",
+    "#55ff55" to "green",
+    "#55ffff" to "aqua",
+    "#ff5555" to "red",
+    "#ff55ff" to "light_purple",
+    "#ffff55" to "yellow",
+    "#ffffff" to "white",
+)
+
+public data class StyleTag(val beforeCursor: String, val placeholder: String, val beforeSel: String, val afterSel: String) {
+    public constructor(opening: String, closing: String) : this("", "", opening, closing)
+    public constructor(simple: String) : this("", "", "<$simple>", "</$simple>")
+}
+
+public val STYLE_WRAPPERS: Map<String, StyleTag> = mapOf(
+    "bold" to StyleTag("b"),
+    "italic" to StyleTag("i"),
+    "underline" to StyleTag("u"),
+    "strikethrough" to StyleTag("st"),
+    "open-url" to StyleTag("<click:open_url:\'", "url", "\'>", "</click>"),
+    "run-command" to StyleTag("<click:run_command:\'", "/command", "\'>", "</click>"),
+    "suggest-command" to StyleTag("<click:suggest_command:\'", "/command", "\'>", "</click>"),
+    "hover-text" to StyleTag("<hover:show_text:\'", "text", "\'>", "</hover>"),
+)
+
+public fun installStyleButtons() {
+    val inputBox = document.getElementById("input")!!.unsafeCast<HTMLTextAreaElement>()
+
+    STYLE_WRAPPERS.forEach { (buttonName, tag) ->
+        val button = document.getElementById("editor-$buttonName-button")!!.unsafeCast<HTMLButtonElement>()
+        button.addEventListener(
+            "click",
+            {
+                handleStyleButton(inputBox, tag)
+            }
+        )
+    }
+
+    document.getElementsByClassName("dropdown-trigger").asList().forEach { element ->
+        element.addEventListener(
+            "click",
+            {
+                // This should hopefully make it so any text selected before pressing the color dropdown should stay visually selected
+                inputBox.focus()
+                element.parentElement!!.classList.toggle("is-active")
+            }
+        )
+    }
+
+    val previewSwatch = document.getElementById("preview-swatch")!!.unsafeCast<HTMLDivElement>()
+    val previewHex = document.getElementById("preview-hex")!!.unsafeCast<HTMLInputElement>()
+
+    // I've given up on kotlin type safety
+    val colorPicker = js("new iro.ColorPicker('#picker')").unsafeCast<ColorPicker>()
```

Comment:
Truly beautiful 😅

Reply:
I tried as hard as i could to write that using external interfaces and shit, but i couldn't figure it out. Since i'm tricking kotlin with js libraries it can't even see anyway, this has about the same effect?